### ISSUE TITLE
remix spa mode cannot with `rollup@4.16.0` , when run `npm run preivew`, it's a empty page

### ISSUE THREAD

**Issue by zouhangwithsweet** [remix-run/remix#9271]
### Reproduction

https://stackblitz.com/edit/remix-run-remix-vuafvw?file=vite.config.ts,app%2Froot.tsx,app%2Froutes%2F_index.tsx,app%2Fentry.client.tsx,app%2Fentry.server.tsx,package.json,README.md

```bash
pnpm add rollup -D

pnpm run build

pnpm run preivew
```

### System Info

```shell
System:
    OS: macOS 14.1.2
    CPU: (8) arm64 Apple M1
    Memory: 74.91 MB / 8.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.10.0 - ~/.nvm/versions/node/v20.10.0/bin/node
    Yarn: 1.22.21 - ~/.nvm/versions/node/v20.10.0/bin/yarn
    npm: 10.2.3 - ~/.nvm/versions/node/v20.10.0/bin/npm
    pnpm: 9.0.4 - ~/Library/pnpm/pnpm
    bun: 1.0.0 - ~/.bun/bin/bun
  Browsers:
    Chrome: 124.0.6367.61
    Safari: 17.1.2
  npmPackages:
    @remix-run/dev: ^2.8.1 => 2.8.1 
    @remix-run/node: ^2.8.1 => 2.8.1 
    @remix-run/react: ^2.8.1 => 2.8.1 
    vite: ^5.1.0 => 5.2.10
```

### Used Package Manager

npm

### Expected Behavior

run `npm run preivew`, it's **not** a empty page

### Actual Behavior

run `npm run preivew`, it's  a emty page

**Comment by huijiewei** [comment]
override rollup to 4.15.0, this is rollup issue.

**Comment by brophdawg11** [comment]
yeah this seems to be a bug in rollup 4.16 when it comes to function parameters with default values.

Our `flattenRoutes` function is defined as:

```js
function flattenRoutes<
  RouteObjectType extends AgnosticRouteObject = AgnosticRouteObject
>(
  routes: RouteObjectType[],
  branches: RouteBranch<RouteObjectType>[] = [],
  parentsMeta: RouteMeta<RouteObjectType>[] = [],
  parentPath = ""
): RouteBranch<RouteObjectType>[] {
  ...
}
```

And it gets transpiled down differently in rollup 4.15 versus 4.16, with 4.16 incorrectly blowing away the `branches` parameter value on every invocation:

```js
// rollup 4.15
function flattenRoutes(routes, branches, parentsMeta, parentPath) {
  if (branches === void 0) {
    branches = [];  // ✅ Only sets the default if no value is passed
  }
  if (parentsMeta === void 0) {
    parentsMeta = [];
  }
  if (parentPath === void 0) {
    parentPath = "";
  }
  // ...
}

// rollup 4.16
function flattenRoutes(routes, branches, parentsMeta, parentPath) {
  {
    branches = []; // ❌ Lost the if statement, so it sets the default every time
  }
  if (parentsMeta === void 0) {
    parentsMeta = [];
  }
  if (parentPath === void 0) {
    parentPath = "";
  }
  // ...
}
```

**Comment by brophdawg11** [comment]
For now the proper workaround is to install rollup@4.15, or set it in the `overrides` section of your `package.json`:

```
"overrides": {
    "rollup": "4.15.0"
}
```

**Comment by brophdawg11** [comment]
This should be resolved in rollup 4.16.2: https://github.com/rollup/rollup/releases/tag/v4.16.2
